Frontend support and adjustments of current banned words blocking code to support new implementation (Fixes #12405)

Scottd125 opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #12405 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
The third and final stage for closing #12405 :)

- Provides front-end support for admins to allow/disallow guilds from using banned words when creating/updating guilds
- API changes for when a guild is created or updated to support the newly added `bannedWordsAllowed` field
- Adjustments to the code that prevents banned words from being posted to use the new `bannedWordsAllowed` field instead of the hard-coded guilds
- Removal of the `guildsAllowingBannedWords.js` file
- Adjustment to the banned words test which relied on the hard-coded guilds to the new implementation

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 
9cbaf3a3-1f48-4557-9f0a-4d403df1f3eb

Scottd125 committed:
Added the "Allow banned words" checkbox to create/update guild forms

Added the "Allow banned words" checkbox to guild create/update forms which are only visible to Admins

Scottd125 committed:
Merge remote-tracking branch 'upstream/develop' into banned-words-allowed-stage-three

Scottd125 committed:
Updated create/update group API to include bannedWordsAllowed property

Added logic to set bannedWordsAllowed depending on conditions

Scottd125 committed:
Updated word banning logic to use bannedWordsAllowed guild value instead

Updated word banning logic to use bannedWordsAllowed guild value and deleted hardcoded guild file & import

Scottd125 committed:
Merge remote-tracking branch 'upstream/develop' into banned-words-allowed-stage-three

Scottd125 committed:
Merge remote-tracking branch 'upstream/develop' into banned-words-allowed-stage-three

Scottd125 committed:
Updated banned words test

Scottd125 committed:
Merge remote-tracking branch 'upstream/develop' into banned-words-allowed-stage-three

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -114,6 +114,16 @@ api.createGroup = {
   url: '/groups',
   middlewares: [authWithHeaders()],
   async handler (req, res) {
+    // Once the request body gets sanitized by the group variable
```
@Scottd125 could we keep the ability of setting `bannedWordsAllowed` only in updateGroup and not when creating it?

Scottd125 replied:
The original description of the issue mentioned adding it to both creating and updating the guild. However, I've updated it to just only be within updateGroup and removed it from createGroup

paglias commented on website/client/src/components/groups/groupFormModal.vue:
```diff
@@ -94,6 +94,40 @@ label.custom-control-label(v-once) {{ $t('allowGuildInvitationsFromNonMembers')
+            class="custom-control-label"
+            for="bannedWordsAllowed"
+          >{{ $t('bannedWordsAllowed') }}</label>
+          <div
```
v-once

Scottd125 replied:
Added

paglias commented on website/client/src/components/groups/groupFormModal.vue:
```diff
@@ -94,6 +94,40 @@ label.custom-control-label(v-once) {{ $t('allowGuildInvitationsFromNonMembers')
+              v-html="icons.information"
+            ></div>
+          </div>
+          <b-tooltip
```
v-once

Scottd125 replied:
Added

paglias commented on test/helpers/api-integration/v3/object-generators.js:
```diff
@@ -121,15 +123,16 @@ async function _upgradeToGroupPlan (groupLeader, group) {
   const numberOfInvites = settings.invites || 0;
   const upgradeToGroupPlan = settings.upgradeToGroupPlan || false;
   const { groupDetails } = settings;
+  const { groupDetailsUpdate } = update;
```
is this needed? why can't `groupDetails ` be used?

Scottd125 replied:
Admittedly, the approach I was taking when updating the test was sloppy due to being quite unfamiliar with the code base. After I got more familiar with it, I noticed the approach I was taking was pretty silly. This has all been amended

paglias reviewed (dismissed):
Left some comments @Scottd125 but the direction is looking good, I'd also like to have a couple new tests showing that the `bannedWordsAllowed` setting can only be set by admins, thanks!

Scottd125 committed:
Merge remote-tracking branch 'upstream/develop' into banned-words-allowed-stage-three

Scottd125 committed:
Pull Request amends

Removed the setting of bannedWordsAllowed during group creation
Added v-once to elements
Updated existing test and added tests related to bannedWordsAllowed functionality

Scottd125 committed:
Merge remote-tracking branch 'upstream/develop' into banned-words-allowed-stage-three

Scottd125 committed:
Small amend

Scottd125 committed:
Small amend

Removed leftover code that was left during testing

Scottd125 commented on website/client/src/components/groups/groupFormModal.vue:
```diff
@@ -94,6 +94,46 @@ label.custom-control-label(v-once) {{ $t('allowGuildInvitationsFromNonMembers')
         -->
       </div>
+      <div
+        v-if="isAdmin && workingGroup.id"
```
In relation to the request to only allow the ability to change bannedWordsAllowed within updateGroup. I've included this conditional statement to ensure the banned words section/checkbox is only displayed on the update guild form modal

paglias commented:
Thanks @Scottd125 ! The PR is working well!

@Alys any objection to this being merged?

Alys commented:
Can it go to delta first?

paglias commented:
@Alys it's on delta

Alys commented:
I've tested this on the delta website and it's looking good! I'd be happy for it to go to production.

Thanks @Scottd125 and thanks @paglias for putting it on delta.

Alys commented:
I've changed the labels to "... merge pending"

paglias commented:
thanks @Scottd125 ! noted towards your 2nd tier
